### DISCUSSION TITLE
What is the meaning of "Writing data to the referee of the result yields an undefined state."

### DISCUSSION THREAD

**Discussion by Jack012a** [nlohmann/json#3836] [Q&A] [answered]
Reference: https://json.nlohmann.me/api/basic_json/get_ref/

> Notes[¶](https://json.nlohmann.me/api/basic_json/get_ref/#notes)
> Undefined behavior
> Writing data to the referee of the result yields an undefined state.

**Question**> What does this meaning?

```
#include <iostream>
#include "nlohmann/json.hpp"

using json = nlohmann::json;

int main()
{
    // create a JSON number
    json value = 17;

    // explicitly getting references
    auto r2 = value.get_ref<json::number_integer_t&>();
    r2 = 188;

    // print the values
    std::cout << "r2: " << r2 << std::endl;
}
```

Output:
`r2: 188`

Based on the reference, what is the undefined behavior here?
Since we can use both `value.get_ref<json::number_integer_t&>()` and `value.get_ref<const json::number_integer_t&>()`, I assume we can modify the return value in the non-const reference case.

Thank you

https://godbolt.org/z/MW99vcPdz

**Comment by falbrechtskirchinger** [comment, marked as answer]
While it may work, you should not rely on it. If the library makes changes that break the observed behavior, that's too bad, you were told not to use it.

**Reply by falbrechtskirchinger** [reply]
For what it's worth, I routinely ignore this restriction and believe it should either be lifted or narrowed. Otherwise, the API should disallow obtaining a non-`const` reference.

**Reply by falbrechtskirchinger** [reply]
Oh, and looking at your code, what you're doing is perfectly fine. You're not modifying the referee. Is that really what you meant to show/ask?

**Reply by nlohmann** [reply]
The original feature was discussed in https://github.com/nlohmann/json/issues/91 where also the remark on undefined behavior stems from. I am open to discuss better wording for the warning.

**Reply by Jack012a** [reply]
Let me give another example below. Basically, I want to know what is the undefined behavior
```
#include <iostream>
#include "nlohmann/json.hpp"
using json = nlohmann::json;

int main()
{
    json value = 17;
    // explicitly getting non-const references
    auto& r3 = value.get_ref<json::number_integer_t&>();
    r3 = 1880; // Modify the referred value
    std::cout << "r3: " << r3 << std::endl;
}
```

**Question**> Does this cause undefined behavior?

Thank you

https://godbolt.org/z/3rWa4Wd16

**Reply by falbrechtskirchinger** [reply]
> The original feature was discussed in #91 where also the remark on undefined behavior stems from. I am open to discuss better wording for the warning.

So, I read that as there was no undefined behavior even at the time the feature was implemented, as long as the correct type is supplied?
Of course, the reference becomes invalid if the JSON value is re-assigned.

Today, things are slightly more complicated. With `JSON_DIAGNOSTICS == 1` appending to a `std::vector &` would result in an invalid state, as the parent object would not be set correctly.

Otherwise, if used properly, there is no undefined behavior. Correct?

**Reply by falbrechtskirchinger** [reply]
> **Question**> Does this cause undefined behavior?

If `value.is_number_integer() == true`, then, in my opinion, there is no undefined behavior. At least not today.
Also, if `value` is reassigned, the reference may become invalid.

However, future versions may introduce undefined behavior, since this I clearly documented. (We have `JSON_DIAGNOSTICS` as an example of such a change.)

**Reply by Jack012a** [reply]
The part gets me confused is that the API offers the following two interfaces:

```
get_ref<const json::number_integer_t&>()
get_ref<json::number_integer_t&>()
```

If we are not supposed to change the value, then why the API provides the non-const reference interface.

Thank you

**Reply by falbrechtskirchinger** [reply]
Have a look at this issue and specifically Niels' comment: https://github.com/nlohmann/json/issues/127#issuecomment-145975155

I think we need to clarify the documentation.
